entityloader/initHandler.js

```diff
@@ -0,0 +1,188 @@
+"use strict";
+var crypto = require('crypto') /*, log = require("../lib/logger.js")*/ ;
+var utils = require('../lib/app-utils.js');
+var names = require('../lib/generate.js').names;
+var mathUtils = require('../lib/math-utils.js');
+
+var core, config;
+
+
+
+function generateNick(sNick, next) {
+	var lowBound = 1,
+		upBound = 1;
+	if (!sNick) sNick = names(6);
+	sNick = sNick.toLowerCase();
+
+
+	function checkUser(suggestedNick, attemptC, callback) {
+		var ct = 0,
+			result = true;
+		var trying = suggestedNick;
+
+		if (attemptC) {
+			lowBound = upBound;
+			upBound = 1 << attemptC;
+			trying += mathUtils.random(lowBound, upBound);
+		}
+
+		if (attemptC >= config.nickRetries) return callback(names(6));
+
+		function done(r) {
+			result &= r;
+			if (++ct >= 3) {
+				if (result) {
+					callback(trying);
+				} else {
+					checkUser(suggestedNick, attemptC + 1, callback);
+				}
+			}
+		}
+
+		function checkRoomUser(type, name) {
+			core.emit(type, {
+				ref: name,
+				session: "internal-loader"
+			}, function(err, data) {
+				if (!err && data && data.results && data.results.length > 0) {
+					done(false);
+				} else {
+					done(true);
+				}
+			});
+		}
+		checkRoomUser("getRooms", trying);
+		checkRoomUser("getUsers", trying);
```

Comment:
`getEntities({ ref: ... })` is available so no need to do two queries here.

Overall I this will need to be refactored. This could also be a lot more efficient if prefix matching entity id's (which we were talking about in the context of autocomplete) is available, do getEntities with username*, find the last one and increment.